crypto/evp/keymgmt_lib.c

```diff
@@ -370,3 +370,82 @@ void *evp_keymgmt_util_gen(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
 
     return keydata;
 }
+
+struct export_todata_st {
+    OSSL_PARAM *params;
+};
+
+/*
+ * Some keys have multiple instances, so mark them as used so that they are only
+ * found once.
+ */
+static OSSL_PARAM *locate_used(const OSSL_PARAM *params, const char *key,\
+                               int used[], int used_len)
+{
+    int i;
+    OSSL_PARAM *p = (OSSL_PARAM *)params;
+
+    for (i = 0; p->key != NULL && i < used_len; p++, i++) {
+        if (!used[i] && strcmp(key, p->key) == 0) {
+            used[i] = 1;
+            return p;
+        }
+    }
+    return NULL;
+}
+
+#define OSSL_PARAM_TODATA_MAX 25
+
+static int todata(const OSSL_PARAM src[], void *cbarg)
```

Comment:
I can't say I like this.  Passing the caller's OSSL_PARAM array to the export and filling the fields they want seems more natural than copying everything here.  Still, if the fields aren't known.....

Could an OSSL_PARAM array and its size be passed to the export and it gets filled but not overflowed without this callback being involved?

Reply:
The export is there to do export 'everything'.. I have just hijacked it..
The callback is normally used for the import process (using the params that are built)

Reply:
Maybe a "query all possible params to export", a "clone a param list" and an "export" would simplify things.

Reply:
The normal export is also a import

Reply:
> The normal export is also a import

... depending on whose point of view.  What libcrypto exports, the provider imports, and vice versa

Reply:
I still believe that a query of all parameters makes sense.  I'd work with a query of all parameters and their values, plus a separate free function.